### DISCUSSION TITLE
Types generation

### DISCUSSION THREAD

**Discussion by PierBJX** [webpack/webpack#15031] [General]
For one of my project, I create a common webpack configuration where you can configure or custom some parameters. 
In order to make it easier to customize, I wanted to add some types to my files. However, I miss some types from webpack that I would like to use (for example, `EntryDescription`). Thus, I looked into webpack to make a PR in order to export more interfaces. 

But looking into the code, some questions came to me;
- Why do we convert some property type to raw. For instance, [`WebpackOptions`  interface](https://github.com/webpack/webpack/blob/main/declarations/WebpackOptions.d.ts#L766) is converted to [`Configration` interface](https://github.com/webpack/webpack/blob/main/types.d.ts#L2026), if you look at `amd` property, it is converted from `Amd` to `false | { [index: string]: any }` . The same for `cache`, from `CacheOptions` to `boolean | MemoryCacheOptions | FileCacheOptions`. It is pity to lose these types whereas it is provided in `WebpackOptions.d.ts`.

- Why don't we export all of the interfaces which are in `WebpackOptions.d.ts` ?

Thanks.

**Comment by alexander-akait** [comment]
Because we can do some breaking changes, and it will be problem for developers, you can get many options using `Configuration["cache"]` and use other ts helpers, but yes, some times can be missed

**Reply by PierBJX** [reply]
Which breaking changes can you have?  Indeed, `CacheOptions` is equal to `boolean | MemoryCacheOptions | FileCacheOptions`, so if in the interface it does not change the type.

**Reply by alexander-akait** [reply]
But you can get `CacheOptions` using `Configuration['cache']`

**Reply by PierBJX** [reply]
Yes sure but I am just curious of what kind of breaking changes it can occur to do `cache?: CacheOptions;` or `cache?: boolean | MemoryCacheOptions | FileCacheOptions;`

**Reply by alexander-akait** [reply]
Most likely, there will be no changes in this particular option, but we have many more options

**Comment by PierBJX** [comment]
About the second question, why don't we export all interfaces from `WebpackOptions.d.ts`? Do you think it could be possible to do so?

Another question, why `WebpackOptionsNormalized` exists? What is the difference with `WebpackOptions` (or `Configuration`)? What is the added value? Why not keeping all in one interface?

**Reply by alexander-akait** [reply]
> About the second question, why don't we export all interfaces from WebpackOptions.d.ts? Do you think it could be possible to do so?

Because it’s wrong, even typescript don't exports all types when you generate declarations, also as I described above, some canb e changes, for example `outputFileSystem`, we can add new methods and it will be problem, developers will need to add the new methods too, we try to be safely as we can, but sometime it is impossible, most of options can be get using `Configuration["option"]`

> Another question, why WebpackOptionsNormalized exists? What is the difference with WebpackOptions (or Configuration)?

Because webpack normalizes options and applies default values, https://github.com/webpack/webpack/blob/main/lib/config/normalization.js, so you have raw options (from your configuration) and normalized (after normalization and default values), normalized options can be useful for plugins which take options from `compiler`/`compilation`, for example getting `output.publicPath`, in `webpack.config.js` you should not touch them

**Reply by PierBJX** [reply]
Alright thanks :)

Each provided webpack configuration goes through the `getNormalizedWebpackOptions`?

**Reply by alexander-akait** [reply]
Yes, each configuration

**Reply by PierBJX** [reply]
Therefore, if we want to type to configuration we provide. What should we used? `WebpackOptions` or `WebpackOptionsNormalized`? Indeed, the property `devServer` does not exist in `WebpackOptions`, only in the normalized. However, in the `WebpackOptionsNormalized`, some properties are mandatory whereas most of the time I don't need to fill it (such as `cache`).

**Reply by alexander-akait** [reply]
Types for dev server here https://github.com/webpack/webpack-dev-server/pull/4119, it will be released in near future, we are working on it, you should use `Configuration`, not `WebpackOptions`/`WebpackOptionsNormalized` https://github.com/webpack/webpack/blob/main/types.d.ts#L2026

**Reply by PierBJX** [reply]
So in `Configuration` (or `WebpackOptions`), the property `devServer` will be added?

Since in your MR, you are adding types, will it be possible to add `EntryDescription` and `EntryItem` please?

**Reply by alexander-akait** [reply]
> So in Configuration (or WebpackOptions), the property devServer will be added?

Yes

> Since in your MR, you are adding types, will it be possible to add EntryDescription and EntryItem please?

It is types for dev server, why you need `EntryDescription`/`EntryItem`?

**Reply by PierBJX** [reply]
I need it because I have a method which converts the provided entry into an `EntryObject` containing only `EntryDescription` with a convenient library definition. At some point, I have a method like this:
`toEntryDescription(entryName: string, entryPoint: EntryItem | EntryDescription): EntryDescription;`

**Reply by alexander-akait** [reply]
We have https://github.com/webpack/webpack/blob/main/types.d.ts#L12785, so you can use it for types

**Reply by PierBJX** [reply]
Yes, I know and this is where I import `Configuration`, `EntryObject` but `EntryDescription` and `EntryItem` are not available
> TS2305: Module '"webpack/types"' has no exported member 'EntryDescription'.
> TS2305: Module '"webpack/types"' has no exported member 'EntryItem'.

**Reply by alexander-akait** [reply]
Just use more common type, i.e. `EntryObject`, other types are not exported by design

**Reply by PierBJX** [reply]
The problem is that `EntryObject` is string index signature. So in my case, if I iterate through this object, I can see that it is `EntryItem` or `EntryDescription` but I can not add the type.

**Reply by PierBJX** [reply]
I have a similar issue for example with the library output. Indeed, the type of the property `library` of the `Output` interface is `Library`. This type is equal to `LibraryName | LibraryOptions;`. However, there is no exported member for `LibraryName` or `Library` from `types.d.ts`. Whereas `LibraryOptions` is available.

**Reply by alexander-akait** [reply]
Because we can change these types, please use more common type from `Configuration["library"]`

**Reply by PierBJX** [reply]
Property `Configuration["library"]` does not exist

**Reply by alexander-akait** [reply]
Fixed:

Configuration["output"]["library"]

**Reply by PierBJX** [reply]
I already tried this, and it does not work :/

> TS2339: Property 'library' does not exist on type 'Output | undefined'.

**Reply by alexander-akait** [reply]
Yep, please use `NonNullable<Configuration["output"]>["library"]`

**Reply by PierBJX** [reply]
Well spotted! Thanks :)

About [my previous question](https://github.com/webpack/webpack/discussions/15031#discussioncomment-1853852), do you see any workaround ?

**Reply by alexander-akait** [reply]
Link is not working...

**Reply by PierBJX** [reply]
If you open it in a new tab, it will work. It is because the message is hidden in `previous replies`

**Reply by alexander-akait** [reply]
You should use the same - `NonNullable<Configuration["entry"]>`

**Reply by PierBJX** [reply]
No because the returned type is `string | (() => string | EntryObject | string[] | Promise<EntryStatic>) | EntryObject | string[]`.

I need the type of the string index signature of `EntryObject` => `EntryItem | EntryDescription`.

**Reply by alexander-akait** [reply]
You should use common types, not specified, just use `NonNullable<Configuration["entry"]>`, please use https://discord.com/invite/typescript if you need more information about ts, these types are internal and we don't want to expose them

**Reply by PierBJX** [reply]
Alright I understand. However, in some cases you expose the internal types... 

`Entry` which is equal `string | (() => string | EntryObject | string[] | Promise<EntryStatic>) | EntryObject | string[]`.
- `Entry` => exposed and accessible from `NonNullable<Configuration["entry"]>`
- `EntryObject` => exposed but not accessible (**internal**)
- `EntryItem` => not exposed and not accessible  (**internal**)
- `EntryDescription` => not exposed and not accessible  (**internal**)

The same for the `Library` which is equal `LibraryName | LibraryOptions`.
- `Library` => not exposed but accessible from `NonNullable<Configuration["output"]>["library"]`  (**internal**)
- `LibraryName` => not exposed and not accessible  (**internal**)
- `LibraryOptions` => exposed  (**internal**)


This is why at the end it is a bit confusing.

**Reply by PierBJX** [reply]
Anyway, thank you for your replies !
